[IMAGING-316] Add support for BigTIFF format

gwlucastrig opened the pull request:
These changes add support for the BigTIFF format. They enable the Commons Imaging library to read TIFF files of size larger than supported by a 32-bit address space.  A dedicated test case and a set of BigTIFF sample files are included.

Please see the Commons Imaging JIRA tracker issue 316 at  [https://issues.apache.org/jira/browse/IMAGING-316](https://issues.apache.org/jira/browse/IMAGING-316) for more detail.

The ability to write BigTIFF files will be reserved for a future pull request

gwlucastrig commented:
Do you have a feel for when this PR might get merged or, failing that, rejected?  I've got some other stuff I'm working on and am trying to plan for the next few weeks.

garydgregory commented on src/main/java/org/apache/commons/imaging/common/BinaryFunctions.java:
```diff
@@ -177,6 +177,44 @@ public static int read4Bytes(final String name, final InputStream is,
         return result;
     }
 
+   public static long read8Bytes(final String name,
```
Javadoc?

gwlucastrig replied:
done

garydgregory reviewed (changes requested):
See my 2 comments. In general, I would be really nice if new public and protected elements carried Javadocs. That will help us going forward I think.

gwlucastrig commented:
Will address these comments and update code.

gwlucastrig commented:
I hope I did this right.  Please let me know if it didn't work. 

I am not adept at Git and was unable to find a clear set of instructions for submitting changes in response to a PR review. I'm sure they're out there, but I didn't find them.

garydgregory commented on src/main/java/org/apache/commons/imaging/common/ByteConversions.java:
```diff
@@ -399,6 +399,43 @@ private static int[] toUInt16s(final byte[] bytes, final int offset, final int l
         return result;
     }
 
+     public static long toLong(final byte[] bytes, final ByteOrder byteOrder) {
```
We should provide Javadocs on all new public and protected methods IMO.

garydgregory replied:
Also, it looks like there is an extra space in front of `public`?

garydgregory commented on src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java:
```diff
@@ -234,6 +234,14 @@ public int[] getIntArrayValue() throws ImagingException {
             final int[] numbers = (int[]) o;
             return Arrays.copyOf(numbers, numbers.length);
         }
+        if (o instanceof long[]){
```
The formatting seems off in the block, missing spaces.

gwlucastrig replied:
done

garydgregory commented on src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java:
```diff
@@ -288,6 +296,55 @@ public int getIntValueOrArraySum() throws ImagingException {
         // return -1;
     }
 
+    public long[] getLongArrayValue() throws ImagingException {
```
We should provide Javadocs on all new public and protected methods IMO.

gwlucastrig replied:
done

garydgregory commented on src/main/java/org/apache/commons/imaging/formats/tiff/TiffField.java:
```diff
@@ -288,6 +296,55 @@ public int getIntValueOrArraySum() throws ImagingException {
+        // return null;
+    }
+
+    public long getLongValue() throws ImagingException {
```
We should provide Javadocs on all new public and protected methods IMO.

gwlucastrig replied:
done

garydgregory commented on src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java:
```diff
@@ -455,27 +473,47 @@ private TiffHeader readTiffHeader(final ByteSource byteSource) throws ImagingExc
-        if (tiffVersion != 42) {
-            throw new ImagingException("Unknown TIFF Version: " + tiffVersion);
-        }
+   private TiffHeader readTiffHeader(final InputStream is) throws ImagingException, IOException {
```
Formatting seems odd.

gwlucastrig replied:
done

garydgregory reviewed (changes requested):
Hi @gwlucastrig 
Thank you for your updates. Please see my comments.

gwlucastrig commented:
Added Javadoc and corrected formatting.  In order to support an additional JUnit test, I needed to add a new toBytes() method for longs.  This method wasn't required for reading BigTIFFs, but it was necessary to implement the JUnit test.  It will be useful in the future for any implementation that writes BigTIFFs (the current implementation only reads them).

garydgregory commented on src/main/java/org/apache/commons/imaging/formats/tiff/TiffReader.java:
```diff
@@ -16,27 +16,29 @@
 import org.apache.commons.imaging.ImagingException;
 import org.apache.commons.imaging.bytesource.ByteSource;
 import org.apache.commons.imaging.common.BinaryFileParser;
+import static org.apache.commons.imaging.common.BinaryFunctions.read2Bytes;
```
All statics should be grouped at the top. Not sure if Checkstyle can check for us though.

gwlucastrig replied:
I ran checkstyle before submitting and it didn't complain.  But I definitely agree that the statics should go first. I will modify this and submit a new commit tonight or tomorrow.

gwlucastrig replied:
done

garydgregory reviewed (approved):
LGTM
